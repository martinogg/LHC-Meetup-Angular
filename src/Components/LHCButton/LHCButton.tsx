
import React, { Component, Children } from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';

interface IProps {
    onSelected(): void
    styles?: any
}

export default class LHCButton extends Component<IProps> {

    constructor(props: IProps) {

        super(props);
    }

    render() {

        return <TouchableOpacity style={[localStyles.container, this.props.styles]} onPress={() => this.props.onSelected()}>
            {this.props.children}
        </TouchableOpacity>
    }
}

const localStyles = StyleSheet.create({
    container: {
        margin: 10,
        backgroundColor: '#2980b6',
        paddingVertical: 15
    },
})