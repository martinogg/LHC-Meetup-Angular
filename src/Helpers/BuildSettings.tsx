const jsonkey = {
    LHC: {
        FIREBASEapiKey: "FIREBASEKEY",
        FIREBASEauthDomain: "FIREBASEKEY",
        FIREBASEdatabaseURL: "FIREBASEKEY",
        FIREBASEstorageBucket: "FIREBASEKEY",
    }
};

const build = 'LHC';

class BuildSettings {
    
    public static get(value: string): any {
        const app = jsonkey[build]
        return app[value]
    }

}

export default BuildSettings;