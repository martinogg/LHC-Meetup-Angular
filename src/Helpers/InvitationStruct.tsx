export interface IInvitation {
    from: string,
    to: string,
    reason: string,
    status: InvitationStatus
}

export interface IInvitationFromFirebase {
    id: string,
    invitation: IInvitation
}

export interface IInvitationFromAndTo {
    from: IInvitationFromFirebase[],
    to: IInvitationFromFirebase[]
}

export enum InvitationStatus {
    New = 'New',
    Accepted = 'Accepted',
    Rejected = 'Rejected'
}

export class Invitation {

    public static create(_from: string, _to: string, _reason: string, _status: InvitationStatus): IInvitation {

        const ret = {
            from: _from,
            to: _to,
            reason: _reason,
            status: _status
        }
        return ret
    }
}