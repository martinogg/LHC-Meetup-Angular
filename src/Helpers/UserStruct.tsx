export interface IUserInterest {
    title: string,
    description: string
}

export interface IUser {
    userName: string,
    userLocation: string,
    userContact: string,
    userInterests: IUserInterest[],
}

export interface IUserFromFirebase {
    id: string,
    user: IUser
}

export class User {

    public static create(name: string, location: string, contact: string, interests: IUserInterest[]): IUser {

        const ret: IUser = {
            userName: name,
            userLocation: location,
            userContact: contact,
            userInterests: interests
        }

        return ret
    }
}

export class UserInterest {

    public static create(_title: string, _description: string): IUserInterest {

        const ret: IUserInterest = {
            title: _title,
            description: _description
        }

        return ret
    }
}