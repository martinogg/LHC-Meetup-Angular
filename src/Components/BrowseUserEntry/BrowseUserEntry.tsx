import React, { Component } from 'react';
import { StyleSheet, Text } from 'react-native';
import { IUser } from '../../Helpers/UserStruct'

import LHCButton from '../LHCButton/LHCButton'

interface IProps {
    onSelected(): void,
    user: IUser
}

export default class BrowseUserEntry extends Component<IProps> {

    constructor(props: IProps) {

        super(props);
    }

    render() {

        return <LHCButton styles={localStyles.container} onSelected={() => this.props.onSelected()}>
            <Text>Name: {this.props.user.userName}</Text>
            <Text>Location: {this.props.user.userLocation}</Text>
            <Text>Contact: {this.props.user.userContact}</Text>
            <Text>Interests: {this.props.user.userInterests}</Text>
        </LHCButton>
    }
}

const localStyles = StyleSheet.create({

    container: {
        margin: 5,
        padding: 5,
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        backgroundColor: '#F55C5F'
    }
    
})