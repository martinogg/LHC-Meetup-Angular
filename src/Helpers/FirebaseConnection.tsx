// TODO Refactor this mess and get some tests done

'use strict'
import React from 'react';
import {
  Alert,
} from 'react-native';
import * as firebase from 'firebase';
import 'firebase/firestore';
import BuildSettings from './BuildSettings';

import { IUser, User, IUserFromFirebase, IUserInterest } from './UserStruct'
import { IInvitation, IInvitationFromFirebase, Invitation, IInvitationFromAndTo, InvitationStatus, IInvitationFromAndToWithUserObjects, IInvitationFromFirebaseWithUserObject } from './InvitationStruct'

var instance: FirebaseConnection | null = null;

class FirebaseConnection {
  firebaseApp: firebase.app.App | null = null;

  static getInstance(): FirebaseConnection {
    if (instance) {
      return instance;
    } else {
      instance = new FirebaseConnection();
      instance.initApp();
      return instance;
    }
  }

  initApp() {
    var firebaseConfig = {
      apiKey: BuildSettings.get("FIREBASEapiKey"),
      authDomain: BuildSettings.get("FIREBASEauthDomain"),
      databaseURL: BuildSettings.get("FIREBASEdatabaseURL"),
      storageBucket: BuildSettings.get("FIREBASEstorageBucket"),
      projectId: BuildSettings.get("FIREBASEprojectId")
    };

    if (firebaseConfig.apiKey != "") {
      this.firebaseApp = firebase.initializeApp(firebaseConfig);
    }
  }

  getProperties(obj, count) {
    if (isNaN(count)) count = Infinity
    var keys = []
    for (var it in obj) {
      if (keys.length > count) break;
      keys.push(it);
    }
    return keys;
  }

  public logout(): Promise<void> {

    return new Promise((resolve, reject) => {
      firebase.auth().signOut().then(() => {

        resolve()
      }, (failReason) => {

        Alert.alert('Error: ' + failReason)
        resolve()
      })
    })
  }

  // TODO TEST
  public checkFBVersion(): Promise<boolean> {
    return new Promise<boolean>((resolve, reject) => {

      if (firebase.auth().currentUser == null) {

        reject('User Not Logged in')
        return
      }

      const localVersion: number = BuildSettings.get("FIREBASEVersion")

      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-settings").get()
        .then((snap) => {

          let fbVersion: number = 99999999
          snap.forEach(setting => {

            if (setting.id == 'version') {

              const data = setting.data()
              fbVersion = +data.version
            }

          });

          resolve(localVersion >= fbVersion)
        })
        .catch(function (error) {

          reject(error)
        });
    })
}


  //TODO TEST
  public buildInvitationsWithUserObjects(invitations: IInvitationFromAndTo): Promise < IInvitationFromAndToWithUserObjects > {

  return new Promise<IInvitationFromAndToWithUserObjects>((resolve, reject) => {

    this.allUsers().then((users) => {

      let fromInvitations: IInvitationFromFirebaseWithUserObject[] = []
      let toInvitations: IInvitationFromFirebaseWithUserObject[] = []

      const createNewInvitation: (invitation: IInvitationFromFirebase) => IInvitationFromFirebaseWithUserObject = (invitation) => {

        const newInvitation: IInvitationFromFirebaseWithUserObject = {

          ...invitation,
          fromObject: this.userObjectForID(invitation.invitation.from, users),
          toObject: this.userObjectForID(invitation.invitation.to, users)
        }

        return newInvitation
      }

      invitations.from.forEach((element) => {

        fromInvitations.push(createNewInvitation(element))
      })

      invitations.to.forEach((element) => {

        toInvitations.push(createNewInvitation(element))
      })

      resolve({ from: fromInvitations, to: toInvitations })

    }).catch((error) => {

      reject(error)
    })

  })
}

  private userObjectForID(id: string, users: IUserFromFirebase[]): IUserFromFirebase {

  let ret: IUserFromFirebase | null = null
  users.forEach((element) => {

    if (element.id == id) {

      ret = element
    }
  })

  if (ret == null) {

    ret = { id: id, user: User.create('Unknown User', '', '', []) }
  }

  return ret
}

  public getInvitations(): Promise < IInvitationFromAndTo > {
  // TODO TEST
  return new Promise<IInvitationFromAndTo>((resolve, reject) => {

    if (firebase.auth().currentUser == null) {

      reject('User Not Logged in')
      return
    }

    const currentUser = firebase.auth().currentUser as firebase.User

    firebase.firestore().collection("LHC-Invitations").get()
      .then((snap) => {

        let ret: IInvitationFromFirebase[] = []

        snap.forEach(invitationInList => {

          const id = invitationInList.id
          const data = invitationInList.data()

          if (data) {
            
            const invitation = Invitation.create(data.from, data.to, data.reason, data.status)
            ret.push({ id: id, invitation: invitation })
          }

        });

        resolve({ from: ret.filter((item) => { return item.invitation.from == currentUser.uid }), to: ret.filter((item) => { return item.invitation.to == currentUser.uid }) })
      })
      .catch(function (error) {

        reject(error)
      });
  })

}

  public deleteInvitation(uid: string): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {

      firebase.firestore().collection("LHC-Invitations").doc(uid).delete()
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}


  public updateInvitationResponse(uid: string, response: InvitationStatus): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {

      firebase.firestore().collection("LHC-Invitations").doc(uid).update({ status: response })
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}

  public updateInvitation(uid: string, reason: string): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {

      firebase.firestore().collection("LHC-Invitations").doc(uid).update({ reason })
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}


  public createNewInvitation(invitation: IInvitation): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {

      const newInvitationId = this.dateEpochPlusUID()
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Invitations").doc(newInvitationId).set(invitation)
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}

  public saveUserDetails(user: IUser): Promise < string > {
  // TODO TEST
  return new Promise<string>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Users").doc(currentUser.uid).set({
        name: user.userName,
        location: user.userLocation,
        contact: user.userContact,
        interests: user.userInterests,
      })
        .then(() => {

          resolve()
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}

  private queryMatchesInterest(query: string, interestList: IUserInterest[]): boolean {

  let found = false
  const queryLower = query.toLowerCase()
  interestList.forEach((interest) => {

    if (interest.title.toLowerCase().includes(queryLower)) {

      found = true
    }
    if (interest.description.toLowerCase().includes(queryLower)) {

      found = true
    }
  })

  return found
}

  public searchOtherUsers(query: string): Promise < IUserFromFirebase[] > {
  // TODO TEST
  return new Promise<IUserFromFirebase[]>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Users").get()
        .then((snap) => {

          let ret: IUserFromFirebase[] = []

          snap.forEach(userInList => {

            const id = userInList.id
            const data = userInList.data()
            const matchesInterest = this.queryMatchesInterest(query, data.interests)

            if (data != null && id != currentUser.uid && matchesInterest) {

              const user = User.create(data.name, data.location, data.contact, data.interests)
              ret.push({ id: id, user: user })
            }

          });

          resolve(ret)
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })

}

  public allUsers(): Promise < IUserFromFirebase[] > {
  // TODO TEST
  return new Promise<IUserFromFirebase[]>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Users").get()
        .then((snap) => {

          let ret: IUserFromFirebase[] = []

          snap.forEach(userInList => {

            const id = userInList.id
            const data = userInList.data()

            const user = User.create(data.name, data.location, data.contact, data.interests)
            ret.push({ id: id, user: user })

          });

          resolve(ret)
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })

}

  public loadUserDetails(): Promise < IUser > {
  // TODO TEST
  return new Promise<IUser>((resolve, reject) => {

    if (firebase.auth().currentUser != null) {
      const currentUser = firebase.auth().currentUser as firebase.User

      firebase.firestore().collection("LHC-Users").doc(currentUser.uid).get()
        .then((snap) => {

          const data = snap.data()

          if (data != null) {
            const user = User.create(data.name, data.location, data.contact, data.interests)

            resolve(user)
          } else {
            reject('An Error Occurred')
          }
        })
        .catch(function (error) {

          reject(error)
        });

    } else {

      reject('User Not Logged in')
    }
  })
}

  public login(username: string, password: string): Promise < string > {
  // TODO TEST

  return new Promise((resolve, reject) => {

    firebase.auth().signInWithEmailAndPassword(username, password).then(() => {

      resolve()
    }, (reason) => {

      reject(reason)
    })
  })
}

  public register(username: string, password: string): Promise < string > {
  // TODO TEST

  return new Promise((resolve, reject) => {

    firebase.auth().createUserWithEmailAndPassword(username, password).then(() => {


      // TODO - Shift this out in to separate email verification
      /*
      if (this.isLoggedIn()) {
 
        firebase.auth().currentUser.sendEmailVerification()
      }
      */

      resolve()
    }, (reason) => {
      reject(reason)
    })
  })
}

  public isLoggedIn(): boolean {
  // TODO TEST

  return firebase.auth().currentUser != null
}

  public getCurrentUserID(): string {
  // TODO TEST
  if (firebase.auth().currentUser == null) {
    return 'Error: Not logged In'
  }

  const currentUser = firebase.auth().currentUser as firebase.User
  return currentUser.uid
}

  private dateEpochPlusUID(): string {
  // TODO TEST
  return (new Date).getTime().toString() + '-' + (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1)
}

sendComment(comment: string) {
  // TODO TEST

  if (comment !== '') {

    const date = this.dateEpochPlusUID()
    if (firebase.auth().currentUser != null) {

      const currentUser = firebase.auth().currentUser as firebase.User
      firebase.firestore().collection("LHC-Comments").doc(date).set({
        'comment': comment
      })
    }
  }
}

loginAnon() {
  let strongThis = this;
  if (this.firebaseApp != null) {
    firebase.auth().signInAnonymously().catch((error) => { });
  }
}

// only allows callback if user is actually logged in
_getCurrentUser(onGetCurrentUserFunc) {

  if (this.firebaseApp != null) {

    var currentUser = firebase.auth().currentUser;
    if (currentUser) {

      onGetCurrentUserFunc(currentUser);
    } else {

      // try again just in case user is back online
      this.loginAnon();
    }
  }
}

getLikeWithId(idStr, callback) {
  this._getCurrentUser((currentUser) => {
    var ref = this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("likes").child(idStr);
    ref.once('value', (snap) => {
      var isSet = snap.val() === true;
      callback(isSet);
    });
  });
}

toggleLikeWithId(idStr, callback) {
  this._getCurrentUser((currentUser) => {
    var ref = this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("likes").child(idStr);

    ref.once('value', (snap) => {
      var alreadySet = snap.val() === true;
      var newSet = alreadySet ? null : true;
      ref.set(newSet).then(() => {
        callback(newSet);
      });
    });
  });
}

setViewedWithId(idStr, onCompleteFunc) {
  this._getCurrentUser((currentUser) => {
    var ref = this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("views").child(idStr);

    ref.set(true).then(() => {
      onCompleteFunc();
    });
  });
}

setLikeWithId(idStr, liked, onCompleteFunc) {
  this._getCurrentUser((currentUser) => {
    var ref = this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("likes").child(idStr);

    ref.set(liked ? true : null).then(() => {
      onCompleteFunc();
    });
  });
}

setDownloadedWithId(idStr) {
  this._getCurrentUser((currentUser) => {
    this.firebaseApp.database().ref().child("users").child("" + currentUser.uid).child("actions").child("downloads").child(idStr).set(true);
  });
}

getDbDecks(onCompleteFunc) {

  if (this.firebaseApp != null) {

    const strongThis = this;

    var itemsRef = this.firebaseApp.database().ref().child("data");

    itemsRef.once('value', (snap) => {

      // get children as an array
      var items = [];
      snap.forEach((child) => {
        items.push(child);
      });
      onCompleteFunc(items);
    });
  }
}

getDbDictLikes(onCompleteFunc) {

  if (this.firebaseApp != null) {
    const strongThis = this;

    var itemsRef = this.firebaseApp.database().ref().child("data");

    itemsRef.once('value', (snap) => {

      // get children as an array
      var items = {};
      snap.forEach((child) => {
        var likesCount = 0 + child.val().likes;
        var viewsCount = 0 + child.val().views;
        items["" + child.key] = { likes: likesCount, views: viewsCount };
      });
      onCompleteFunc(items);
    });
  }
}



alert(message) {
  Alert.alert(
    'Error',
    'Error:' + message,
    [
      { text: 'OK', onPress: () => console.log('OK Pressed') },
    ],
    { cancelable: false }
  );
}

}

export default FirebaseConnection // TODO FirebaseConnection is old style, remove this in favor of REDUX Version in the future

export const firebaseService = FirebaseConnection.getInstance().firebaseApp as firebase.app.App