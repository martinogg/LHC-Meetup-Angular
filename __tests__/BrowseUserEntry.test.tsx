import { shallow } from 'enzyme';
import React from 'react';
import 'react-native';

import { User, IUser } from '../src/Helpers/UserStruct'

import LHCButton from '../src/Components/LHCButton/LHCButton'

import BrowseUserEntry from '../src/Components/BrowseUserEntry/BrowseUserEntry'

import renderer from 'react-test-renderer';

test('BrowseUserEntry matches snapshot', () => {

    const user = User.create('a', 'b', 'c', 'd')
    expect(renderer.create(<BrowseUserEntry user={user} onSelected={() => { }} />)).toMatchSnapshot();
})

test('BrowseUserEntry matches snapshot2', () => {

    const user = User.create('e', 'f', 'g', 'h')
    expect(renderer.create(<BrowseUserEntry user={user} onSelected={() => { }} />)).toMatchSnapshot();
})

test('BrowseUserEntry responds to press', () => {

    const user = User.create('e', 'f', 'g', 'h')
    const callback = jest.fn()
    const wrapper = shallow(<BrowseUserEntry user={user} onSelected={() => { callback() }} />);
    const sut: any = wrapper.instance()

    sut.goToEditScreen = jest.fn()
    wrapper.find(LHCButton).first().simulate('selected')

    expect(callback).toHaveBeenCalledTimes(1)
})