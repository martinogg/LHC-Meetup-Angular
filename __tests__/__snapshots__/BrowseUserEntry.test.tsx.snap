// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowseUserEntry matches snapshot 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F55C5F",
      "flex": 1,
      "justifyContent": "center",
      "margin": 5,
      "opacity": 1,
      "padding": 5,
      "paddingVertical": 15,
    }
  }
>
  <Text>
    Name: 
    a
  </Text>
  <Text>
    Location: 
    b
  </Text>
  <Text>
    Contact: 
    c
  </Text>
  <Text>
    Interests: 
    d
  </Text>
</View>
`;

exports[`BrowseUserEntry matches snapshot2 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F55C5F",
      "flex": 1,
      "justifyContent": "center",
      "margin": 5,
      "opacity": 1,
      "padding": 5,
      "paddingVertical": 15,
    }
  }
>
  <Text>
    Name: 
    e
  </Text>
  <Text>
    Location: 
    f
  </Text>
  <Text>
    Contact: 
    g
  </Text>
  <Text>
    Interests: 
    h
  </Text>
</View>
`;
